draw/init_plan9.go

```diff
@@ -0,0 +1,503 @@
+	Windows     *Image
+	DPI         int
+
+	firstfont *Font
+	lastfont  *Font
+
+	White       *Image // Pre-allocated color.
+	Black       *Image // Pre-allocated color.
+	Opaque      *Image // Pre-allocated color.
+	Transparent *Image // Pre-allocated color.
+
+	DefaultFont    *Font
+	DefaultSubfont *Subfont
+
+	conn        *drawFile
+	ctl         *os.File
+	mousectl    *Mousectl
+	keyboardctl *Keyboardctl
+
+	mtpt string
+}
+
+// An Image represents an image on the server, possibly visible on the display.
+type Image struct {
+	Display *Display
+	id      uint32
+	Pix     Pix             // The pixel format for the image.
+	Depth   int             // The depth of the pixels in bits.
+	Repl    bool            // Whether the image is replicated (tiles the rectangle).
+	R       image.Rectangle // The extent of the image.
+	Clipr   image.Rectangle // The clip region.
+	Origin  image.Point     // Of image in screen, for mouse warping.
+	next    *Image
+	Screen  *Screen // If non-nil, the associated screen; this is a window.
+}
+
+// A Screen is a collection of windows that are visible on an image.
+type Screen struct {
+	Display *Display // Display connected to the server.
+	id      uint32
+	Fill    *Image // Background image behind the windows.
+}
+
+// Refresh algorithms to execute when a window is resized or uncovered.
+// Refmesg is almost always the correct one to use.
+const (
+	Refbackup = 0
+	Refnone   = 1
+	Refmesg   = 2
+)
+
+const deffontname = "*default*"
+
+// Init starts and connects to a server and returns a Display structure through
+// which all graphics will be mediated. The arguments are an error channel on
+// which to deliver errors (currently unused), the name of the font to use (the
+// empty string may be used to represent the default font), the window label,
+// and the window size as a string in the form XxY, as in "1000x500"; the units
+// are pixels.
+func Init(errch chan<- error, fontname, label, winsize string) (*Display, error) {
+	if errch == nil {
+		errch = make(chan error, 1)
+	}
+
+	var err error
+
+	d := &Display{
+		errch: errch,
+	}
+
+	// Lock Display so we maintain the contract within this library.
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	if dbg := os.Getenv("DRAWDEBUG"); dbg != "" {
+		d.debug = true
+	}
+
+	width, height := 800, 600
+	if winsize != "" {
+		t := strings.Split(winsize, "x")
+		if len(t) != 2 {
+			return nil, fmt.Errorf("bad winsize, must be $widthx$height")
+		}
+		width, err = strconv.Atoi(t[0])
+		if err != nil {
+			return nil, fmt.Errorf("bad width in winsize: %s", err)
+		}
+		height, err = strconv.Atoi(t[1])
+		if err != nil {
+			return nil, fmt.Errorf("bad height in winsize: %s", err)
+		}
+	}
+
+	wsys := os.Getenv("wsys")
+	if wsys == "" {
+		return nil, fmt.Errorf("$wsys not set")
+	}
+	wsysfd, err := os.OpenFile(wsys, os.O_RDWR, 0666)
+	if err != nil {
+		return nil, err
+	}
+	// note: must not close wsysfd, or fd's get mixed up...
+	d.mtpt = "/n/duit." + path.Base(wsys)
```

Comment:
what's this? an artifact left over from testing/porting?

Reply:
sorry, i didn't see the commit message before i wrote this.

why don't you follow plan9's newwindow and use `/mnt/wsys` with a bind over `/dev`?

Reply:
i tried an rfork(RFNAMEG) briefly (like in newwindow), but it seemed to give me trouble with multiple go procs having different namespace.  perhaps it's needed earlier in application setup, and only once, but i didn't want to go there (yet).
also i don't bind on (before) /dev because i want to make it possible for one application to have multiple windows (like in examples/alert).
i'm always using $wsys and a new window so writes to stdout/stderr don't clobber the ui. useful for me, at least in this stage. perhaps a first window should take over the window at /dev, and later windows should use $wsys...
i'm currently doing a mount on /n/ because mntgen is running there. if there's a way to create & access a window without a new mount, i'm interested, sounds easier. /mnt/wsys/wsys seems to be getting filled with new windows, but don't know what the contract is there.

btw, i've removed the build tags for plan9.

thanks for reviewing my changes.

Reply:
yes, process specific state such as mount in go's multiprocess model on plan 9 can be a problem, similar to setuid/threads problem on unix.

i think the best thing to do would be give users the tools and documentation to solve this as they see fit. this would probably mean allowing a custom mount path to be passed, so you can specify that you want /mnt/wsys (with the caller doing rfork RFNAMEG) or /n/$id (without RFNAMEG).

in the former case, it would probably be good to document that if rfork RFNAMEG is used, `runtime.LockOSThread` must also be used, and operations on that window must happen inside that goroutine.

really, anyone intimate with plan9 and go would know about these problems. but it can't hurt to document it.

side note - there's no mntgen on plan 9 from bell labs. it's a 9front invention.